projects/snapshot_bot/snapshot_bot.cc

```diff
@@ -311,98 +326,100 @@ class RobotFSM : FSM<State>::StateHandler
-
-                    // If test duration is longer than motor command interval, this schedule cannot be maintained so give a warning
-                    if(m_TestDuration > m_Config.getMotorCommandInterval()) {
-                        LOGW << "last test took " << m_TestDuration.count() << "ms - this is longer than desired motor command interval (" << m_Config.getMotorCommandInterval().count() << "ms)";
-                    }
+                    // Write extra logging data
+                    m_LogFile << ", " << objectData.getFrameNumber() << ", " << position[0] << ", " << position[1] << ", " << position[2] << ", " << attitude[0] << ", " << attitude[1] << ", " << attitude[2];
+                }
 
-                    // Reset move time
-                    m_LastMotorCommandTime = motorTime;
+                // If we should save diagnostics when testing
+                if(m_Config.shouldSaveTestingDiagnostic()) {
+                    const std::string filename = "test" + m_Config.getTestingSuffix() + "_" + std::to_string(m_TestImageIndex++) + ".png";
+                    m_LogFile << ", " << filename;
+                    // Build path to test image and save
+                    const auto testImagePath = m_Config.getOutputPath() / filename;
+                    cv::imwrite(testImagePath.str(), m_Unwrapped);
+                }
 
-                    // Determine how fast we should turn based on the absolute angle
-                    auto turnSpeed = m_Config.getTurnSpeed(m_Memory->getBestHeading());
+                m_LogFile << std::endl;
 
-                    // If we should turn, do so
-                    if(turnSpeed > 0.0f) {
-                        const float motorTurn = (m_Memory->getBestHeading() <  0.0_deg) ? turnSpeed : -turnSpeed;
-                        m_Motor.tank(motorTurn, -motorTurn);
+                // If we should stream output
+                /*if(m_Config.shouldStreamOutput()) {
+                    // Attempt to dynamic cast memory to a perfect memory
+                    PerfectMemory *perfectMemory = dynamic_cast<PerfectMemory*>(m_Memory.get());
+                    if(perfectMemory != nullptr) {
+                        // Get matched snapshot
+                        const cv::Mat &matchedSnapshot = perfectMemory->getBestSnapshot();
+
+                        // Calculate difference image
+                        cv::absdiff(matchedSnapshot, m_Unwrapped, m_DifferenceImage);
+
+                        char status[255];
+                        sprintf(status, "Angle:%f deg, Min difference:%f", degree_t(perfectMemory->getBestHeading()).value(), perfectMemory->getLowestDifference());
+                        cv::putText(m_DifferenceImage, status, cv::Point(0, m_Config.getUnwrapRes().height -20),
+                                    cv::FONT_HERSHEY_COMPLEX_SMALL, 1.0, 0xFF);
+
+                        // Send annotated difference image
+                        //m_NetSink.sendFrame(m_DifferenceImage);
```

Comment:
Does the networking code work uncommented? I seem to remember that the only thing broken was that there was a change to the API for ``Net::Server`` so now you need to call ``waitForConnection()`` and use the ``Net::Connection`` object returned to send messages.

Alternatively you could delete this stuff if you don't want it.

Reply:
I will have a go at re-enabling it

Reply:
Oh yeah right - I remember what the problem was now. Whether streaming is enabled or not is controllable at runtime so, as a ``Connection`` is required to make a ``NetSink``, I'd need to switch to using a ``std::unique_ptr<NetSink>`` which is fine *but*  ``Server::waitForConnection`` returns an un-copyable ``Connection`` which I can't create a unique pointer from...what are your thoughts?

Reply:
I wrote a move constructor for ``Net::Connection``s (though in retrospect I should probably just have returned a ``std::unique_ptr`` from ``waitForConnection()`` instead), so you should be able to do this:
```c++
auto p = std::make_unique<Net::Connection>(server.waitForConnection());
```

Reply:
Woah! I had no idea you could move construct unique pointers from stack-allocated memory :astonished: